node/src/utils.rs

```diff
@@ -811,3 +817,400 @@ pub async fn get_graph(
     let res: Bitvm2Graph = serde_json::from_str(graph.raw_data.unwrap().as_str())?;
     Ok(res)
 }
+
+pub async fn publish_graph_to_ipfs(
+    client: &BitVM2Client,
+    graph_id: Uuid,
+    graph: &Bitvm2Graph,
+) -> Result<String, Box<dyn std::error::Error>> {
+    fn write_tx(
+        base_dir: &str,
+        tx_name: IpfsTxName,
+        tx: &Transaction,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        // write tx_hex to base_dir/tx_name
+        let tx_hex = serialize_hex(tx);
+        let tx_cache_path = format!("{}{}", base_dir, tx_name.as_str());
+        let mut file = File::create(&tx_cache_path)?;
+        file.write_all(tx_hex.as_bytes())?;
+        Ok(())
+    }
+
+    let base_dir = format!("{}{}/", IPFS_GRAPH_CACHE_DIR, graph_id.to_string());
+    fs::create_dir_all(base_dir.clone())?;
+    write_tx(&base_dir, IpfsTxName::AssertCommit0, graph.assert_commit.commit_txns[0].tx())?;
+    write_tx(&base_dir, IpfsTxName::AssertCommit1, graph.assert_commit.commit_txns[1].tx())?;
+    write_tx(&base_dir, IpfsTxName::AssertCommit2, graph.assert_commit.commit_txns[2].tx())?;
+    write_tx(&base_dir, IpfsTxName::AssertCommit3, graph.assert_commit.commit_txns[3].tx())?;
+    write_tx(&base_dir, IpfsTxName::AssertInit, graph.assert_init.tx())?;
+    write_tx(&base_dir, IpfsTxName::AssertFinal, graph.assert_final.tx())?;
+    write_tx(&base_dir, IpfsTxName::Challenge, graph.challenge.tx())?;
+    write_tx(&base_dir, IpfsTxName::Disprove, graph.disprove.tx())?;
+    write_tx(&base_dir, IpfsTxName::Kickoff, graph.kickoff.tx())?;
+    write_tx(&base_dir, IpfsTxName::Pegin, graph.pegin.tx())?;
+    write_tx(&base_dir, IpfsTxName::Take1, graph.take1.tx())?;
+    write_tx(&base_dir, IpfsTxName::Take2, graph.take2.tx())?;
+    let cids = client.ipfs.add(&Path::new(&base_dir)).await?;
+    let _ = fs::remove_dir_all(base_dir);
```

Comment:
we should return the error when `remove` fails.  just use `fs::remove_dir_all(base_dir)?;`.

Reply:
> we should return the error when `remove` fails. just use `fs::remove_dir_all(base_dir)?;`.

the existence of a cached file won't affect subsequent execution, is it necessary to return a failure?